### PR TITLE
New `InstanceOfPatternMatch` recipe

### PR DESCRIPTION
**Pull request by knutwannheden**
Implement a recipe which for Java 17+ transforms `instanceof` expressions which have a corresponding type cast within their flow scope (see https://openjdk.org/jeps/394 for details) to an `instanceof` with a pattern variable and then also substitutes the type cast expressions accordingly.

### CODE REVIEW

**Comment by knutwannheden** [comment]
This is still a draft. There are still a few TODOs and the code is probably not very OpenRewrite-idiomatic, as I probably missed out on some helper utilities or misunderstood how to use all available constructs. A critical review would be most appreciated!

File: rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    @Data
+    private static class InstanceOfPatternReplacements {
+        private final J root;
+        private final Map<Expression, J.InstanceOf> instanceOfs = new HashMap<>();
+        private final Map<J.InstanceOf, Set<J>> contexts = new HashMap<>();
+        private final Map<J.TypeCast, J.InstanceOf> replacements = new HashMap<>();
+
+        public void register(J.InstanceOf instanceOf, Set<J> contexts) {
+            Expression expression = instanceOf.getExpression();
+            Optional<Expression> existing = instanceOfs.keySet().stream().filter(e -> SemanticallyEqual.areEqual(e, expression)).findAny();
```
Comment (knutwannheden) [inline rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:144]: Whether all semantically equal expressions or only simple variable references should be refactored by this recipe or not should probably be controlled by an option, as more complex expressions could have side effects. But I think that would be a very weird coding style...

**Comment by knutwannheden** [comment]
An interesting enhancement could be to inline the expression, if it isn't referenced anywhere else. E.g.
```java
Object o = foo.bar();
if (o instanceof String && ((String) o).isEmpty()) {
    // ...
}
```
would become:
```java
if (foo.bar() instanceof String o && o.isEmpty()) {
    // ...
}
```
if `o` isn't referenced anywhere outside the `if`-statement. I would however propose to file a separate enhancement request for this and maybe that should even be a completely independent (reusable) visitor.

**Comment by knutwannheden** [comment]
@sambsnyd I consider this PR to be ready to be reviewed now. I still think that some things may be a bit non-idiomatic and would be glad for some tips on how to improve the code. (I am for instance wondering if it would be more idiomatic to schedule a separate after-recipe for each occurrence, rather than try to collect them all into a single input for the visitor.)

To be on the safe side, I currently exclude any expressions, which may have side effects (as determined by `Expression#getSideEffects()`). I noticed that a few expression types were missing the implementation for this method or had an incorrect implementation. I will look into this in a separate PR. (I would also like to consider support for IDEA's `@Contract(pure=true)` and possibly also the Checker framework's corresponding annotation, but also that will be a separate issue.)

File: rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java (hunk 1 above)
Comment (sambsnyd) [reply rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:159]: Without saying that this usage is problematic or anything, note that `SemanticallyEqual` is not really comprehensive or even entirely well defined. There are lots of things that could be considered equal which are not implemented.

File: rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+        return VariableNameUtils.generateVariableName(shortVariableBaseName((TypeTree) instanceOf.getClazz()), cursor, INCREMENT_NUMBER);
+    }
+
+    private static String shortVariableBaseName(TypeTree typeTree) {
+        String baseName = shortVariableBaseName(typeTree.getType());
+        return typeTree instanceof J.ArrayType ? baseName + 's' : baseName;
+    }
+
+    private static String shortVariableBaseName(@Nullable JavaType type) {
+        // the instanceof operator only accepts classes (without generics) and arrays
```
Comment (sambsnyd) [inline rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:239]: Doesn't look like this handles arrays of arrays. e.g.: `String[][]`. Not sure how common it is that we would run into that.

**Review by sambsnyd — APPROVED** [review]
Nice work. The recipe is complex, but the operation being performed is complex, so I don't see an obvious way to simplify things.

**Comment by sambsnyd** [comment]
> @sambsnyd I consider this PR to be ready to be reviewed now. I still think that some things may be a bit non-idiomatic and would be glad for some tips on how to improve the code. (I am for instance wondering if it would be more idiomatic to schedule a separate after-recipe for each occurrence, rather than try to collect them all into a single input for the visitor or possibly it is even considered better practice to directly "inline" the visitor into the respective `visitXyz()` method and call the `visit(Tree, P, Cursor)` method on it.)

Collecting them all into a single input for the visitor is definitely better for performance than separate after-recipe for each. Better to pay the cost of traversing the LST once rather than `n` times.
I have developed a preference for `visitXyz()` rather than `doNext()`. I often find it easier to debug and reason about the state of the LST when the change I'm looking to effect is made immediately. There is nothing wrong with using `doNext()`, and it is particularly convenient when the change you are trying to effect is going to happen elsewhere in the tree.

> To be on the safe side, I currently exclude any expressions, which may have side effects (as determined by `Expression#getSideEffects()`). I noticed that a few expression types were missing the implementation for this method or had an incorrect implementation. I will look into this in a separate PR. (I would also like to consider support for IDEA's `@Contract(pure=true)` and possibly also the Checker framework's corresponding annotation, but also that will be a separate issue.)

Those sound like good improvements

**Comment by knutwannheden** [comment]
> Nice work. The recipe is complex, but the operation being performed is complex, so I don't see an obvious way to simplify things.

I did figure out a way to slightly simplify things. Since I already used the cursor for the containing statement to "communicate" between the `instanceof` and the type cast expressions, I now added a `postVisit()` method which simply checks for the corresponding message in the cursor and if it finds a non-empty result, it means that there is a replacement to be performed and it can post-process the statement using the other visitor. This way the second visitor only needs to process small sub-trees rather than the whole source again.

I was looking for a mechanism to instruct a cursor to perform some post-processing action, once the visitor leaves that scope. As it turns out the `postVisit()` method can be used for this purpose when adding a message to the corresponding cursor.

File: rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java (hunk 2 above)
Comment (knutwannheden) [reply rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:242]: Good point. Let me add a test for this and see if it needs fixing.

Comment (knutwannheden) [reply rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:242]: For a multi-dimensional array like `int[][]` the LST node is an `J.ArrayType` with `elementType = int` and `dimension` a list of size `2`. It appears to work correctly. Added a test case for this.

**Comment by knutwannheden** [comment]
@sambsnyd Thanks for your review feedback! I did find and fix one more bug and added a few more tests. There are probably more corner cases which should be tested, but I can't think of them right now.

File: rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+    @Nullable
+    private static JavaType toJavaType(TypedTree typeTree) {
+        if (typeTree instanceof J.ArrayType) {
+            JavaType.Array result = new JavaType.Array(null, ((J.ArrayType) typeTree).getElementType().getType());
+            for (int i = 0; i < ((J.ArrayType) typeTree).getDimensions().size() - 1; i++) {
+                result = new JavaType.Array(null, result);
+            }
+            return result;
+        }
+        return typeTree.getType();
```
Comment (knutwannheden) [inline rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:256]: @sambsnyd For `J.ArrayType` the `TypedTree#getType()` implementation always returns the element type (e.g. `int` for `int[][]`). Intuitively I would have expected the returned `JavaType` to also represent `int[][]`. But assuming it is implemented correctly, I implemented my own conversion here, so that multi-dimensional array types get compared correctly in the context of this recipe.

Comment (sambsnyd) [reply rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:256]: `J.ArrayType` represents an appearance of a type in code, whereas `JavaType.Array` is metadata about that type. 
Consider this variable declaration which includes both:
`String[] s;`

There is a `J.ArrayType` for the "String[]" part of the declaration, which contains the whitespace and markers and such and is rendered in text. If you were to call `J.Identifier.getType()` on that identifier "s" it would return a `JavaType.Array` representing an array of Strings. `JavaType.Array` is metadata and does not literally appear in the text

I agree that this is confusing because these are obviously similar concepts. If you were to `null` out the type information, the `JavaType.Array`, you'd still get the same text printed out in the end. If you were to `null` out the `J.ArrayType` you'd print out the ` s;`, which is obviously not valid Java. Hopefully that helps to clarify the relationship.

Comment (sambsnyd) [inline rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:256]: Why is this whole method's implementation not just `typeTree.getType()`? I don't understand the processing you are doing on `typeTree`. I would expect that `J.ArrayType` for `String[][]` would return a corresponding `JavaType.Array` without requiring any post-processing. Is that not the case?

Comment (knutwannheden) [reply rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:256]: Correct. The `getType()` method returns the array element type (e.g. `int`) for a `J.ArrayType`. that is what confused me all along and I intuitively suspected this to be wrong.

Comment (knutwannheden) [reply rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:256]: That part is quite clear to me. That is no different than in other ASTs I've been working with. Well, there is one difference: Normally the AST wouldn't already contain the types, since the AST was produced by a parser, rather than a compiler, as in this case.

Sorry about not explaining my question more clearly.

Comment (knutwannheden) [reply rewrite-java/src/main/java/org/openrewrite/java/InstanceOfPatternMatch.java:257]: This question will be addressed with #2713. In the scope of that issue this method can then hopefully be removed again.

**Comment by timtebeek** [comment]
[Over on this issue](https://github.com/openrewrite/rewrite-migrate-java/issues/176#issuecomment-1406446723) the question came up if Java 9+ recipes should go into rewrite-migrate-java rather than openrewrite/rewrite.

Intuitively I'd think any "new"-ish Java features would best fit into `org.openrewrite.java.migrate`, and be aggregated in that module in yaml migration recipes to complete a migration. That way there's also no question what Java source level is required for openrewrite/rewrite: 8 (for now), to best fit with what we're seeing in the wild.

What are your thoughts on that @sambsnyd ?

**Comment by sambsnyd** [comment]
@timtebeek I see rewrite-migrate-java as being for recipes about migrating to new versions of java. I don't think we _need_ to put all recipes applicable only to newer versions of java there. This feels like a cleanup recipe more than a migration recipe, so putting it into rewrite-java feels reasonable to me

**Comment by knutwannheden** [comment]
> This feels like a cleanup recipe more than a migration recipe, so putting it into rewrite-java feels reasonable to me

@sambsnyd Should the recipe be moved to the `cleanup` subpackage?
